src/Microsoft.IdentityModel.Protocols.PoP/SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1231 @@
+        /// </summary>
+        /// <param name="jkuSetUrl">A URL reference to a PoP JWK set.</param>
+        /// <param name="signedHttpRequestValidationData">A structure that wraps parameters needed for SignedHttpRequest validation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A resolved PoP <see cref="SecurityKey"/>.</returns>
+        protected virtual async Task<SecurityKey> ResolvePopKeyFromJkuAsync(string jkuSetUrl, SignedHttpRequestValidationData signedHttpRequestValidationData, CancellationToken cancellationToken)
+        {
+            var popKeys = await GetPopKeysFromJkuAsync(jkuSetUrl, signedHttpRequestValidationData, cancellationToken).ConfigureAwait(false);
+            var popKeyCount = popKeys.Count;
+
+            if (popKeyCount == 0)
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidPopKeyException(LogHelper.FormatInvariant(LogMessages.IDX23020, popKeyCount.ToString())));
+            else if (popKeyCount > 1)
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidPopKeyException(LogHelper.FormatInvariant(LogMessages.IDX23020, popKeyCount.ToString())));
+            else
+                return popKeys[0];
+        }
+
+        /// <summary>
+        /// Resolves a PoP <see cref="SecurityKey"/> from the URL reference to a PoP key.  
+        /// </summary>
+        /// <param name="jkuSetUrl">A URL reference to a PoP JWK set.</param>
+        /// <param name="kid">A PoP key identifier.</param>
+        /// <param name="signedHttpRequestValidationData">A structure that wraps parameters needed for SignedHttpRequest validation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A resolved PoP <see cref="SecurityKey"/>.</returns>
+        protected virtual async Task<SecurityKey> ResolvePopKeyFromJkuAsync(string jkuSetUrl, string kid, SignedHttpRequestValidationData signedHttpRequestValidationData, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(kid))
+                throw LogHelper.LogArgumentNullException(nameof(kid));
+
+            var popKeys = await GetPopKeysFromJkuAsync(jkuSetUrl, signedHttpRequestValidationData, cancellationToken).ConfigureAwait(false);
+
+            foreach (var key in popKeys)
+            {
+                if (string.Equals(key.KeyId, kid.ToString(), StringComparison.Ordinal))
+                    return key;
+            }
+
+            throw LogHelper.LogExceptionMessage(new SignedHttpRequestInvalidPopKeyException(LogHelper.FormatInvariant(LogMessages.IDX23021, kid, string.Join(", ", popKeys.Select(x => x.KeyId ?? "Null")))));
+        }
+
+        /// <summary>
+        /// Gets a JWK set of PoP <see cref="SecurityKey"/> from the <paramref name="jkuSetUrl"/>.
+        /// </summary>
```

Comment:
"Remove repeated query params according to the spec:"

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1257/commits/27eed45681bef30e04366eff75ecb1a35e26e784.